File: ruby/get-filters.rb
#
# List all Gmail filters and display their criteria and actions.
#
# run:
# ruby get-filters.rb | tee get-filters.out
#
require 'fileutils'
require_relative 'authorize.rb'
require_relative 'labels.rb'

# Initialize the API
APPLICATION_NAME = 'Gmail API Ruby Quickstart'.freeze
# https://www.rubydoc.info/github/google/google-api-ruby-client/Google/Apis/GmailV1
# https://www.rubydoc.info/github/google/google-api-ruby-client/Google/Apis/GmailV1/GmailService
service = Google::Apis::GmailV1::GmailService.new
service.client_options.application_name = APPLICATION_NAME
service.authorization = authorize

user_id = 'me'
labelsNameMap, labelsIdMap = getLabelsMaps(service, user_id)

# https://www.rubydoc.info/github/google/google-api-ruby-client/Google/Apis/GmailV1/GmailService#list_user_setting_filters-instance_method
# https://www.rubydoc.info/github/google/google-api-ruby-client/Google/Apis/GmailV1/ListFiltersResponse
puts "--------------------------------------------------------------------------------"
result = service.list_user_setting_filters(user_id)
puts 'None found' if result.filter.empty?
puts "result.filter.count=[#{result.filter.count}]"

result.filter.each { |filter| 
  # https://www.rubydoc.info/github/google/google-api-ruby-client/Google/Apis/GmailV1/Filter
  puts "============================================================"
  puts "filter.id=[#{filter.id}]"
  # https://www.rubydoc.info/github/google/google-api-ruby-client/Google/Apis/GmailV1/FilterCriteria
  puts "filter.criteria.from=[#{filter.criteria.from}]"
  # https://www.rubydoc.info/github/google/google-api-ruby-client/Google/Apis/GmailV1/FilterAction
  puts "filter.action.add_label_ids=[#{filter.action.add_label_ids}]"
#  filter.action.add_label_ids.each { |label_id|
#    puts "  label_id=[#{label_id}] label_name=[#{labelsIdMap[label_id]}]"
#  }
  puts "filter.action.remove_label_ids=[#{filter.action.remove_label_ids}]"
  puts "filter.action.forward=[#{filter.action.forward}]"
}

File: python/get_filters.py
#!/usr/bin/env python3
#
# List all Gmail filters and display their criteria and actions.
#
# run:
# python get_filters.py | tee get-filters.out
#
from googleapiclient.discovery import build
from authorize import authorize
from labels import get_labels_maps

# Initialize the API
APPLICATION_NAME = 'Gmail API Python Quickstart'
service = build('gmail', 'v1', credentials=authorize())

user_id = 'me'
labels_name_map, labels_id_map = get_labels_maps(service, user_id)

print("--------------------------------------------------------------------------------")
results = service.users().settings().filters().list(userId=user_id).execute()
filters = results.get('filter', [])

if not filters:
    print('None found')
else:
    print(f"result.filter.count=[{len(filters)}]")

for filter_obj in filters:
    print("============================================================")
    filter_id = filter_obj.get('id', '')
    criteria = filter_obj.get('criteria', {})
    action = filter_obj.get('action', {})
    
    from_addr = criteria.get('from', '')
    add_label_ids = action.get('addLabelIds', [])
    remove_label_ids = action.get('removeLabelIds', [])
    forward = action.get('forward', '')
    
    print(f"filter.id=[{filter_id}]")
    print(f"filter.criteria.from=[{from_addr}]")
    print(f"filter.action.add_label_ids=[{add_label_ids}]")
    # Uncomment to see label names:
    # for label_id in add_label_ids:
    #     label_name = labels_id_map.get(label_id, '')
    #     print(f"  label_id=[{label_id}] label_name=[{label_name}]")
    print(f"filter.action.remove_label_ids=[{remove_label_ids}]")
    print(f"filter.action.forward=[{forward}]")
